File: agent/tools.py
"""Tool definitions for the RC car agent."""

# Serial protocol byte IDs
# Commands (sent to Pi):
#   0x01 = turn left 15°
#   0x02 = turn right 15°
#   0x03 = move forward 1ft
#   0x04 = move forward 2in
#   0x05 = move backward 1ft
#   0x06 = move backward 2in
# Responses (from Pi):
#   0xAA = ACK (command received)
#   0xBB = DONE (movement complete)
#   0xFF = ERROR

CMD_TURN_LEFT_15 = 0x01
CMD_TURN_RIGHT_15 = 0x02
CMD_MOVE_FWD_1FT = 0x03
CMD_MOVE_FWD_2IN = 0x04
CMD_MOVE_BWD_1FT = 0x05
CMD_MOVE_BWD_2IN = 0x06

RESP_ACK = 0xAA
RESP_DONE = 0xBB
RESP_ERROR = 0xFF

TOOL_DEFS = [
    {
        "name": "turn_left_15",
        "description": "Turn the car left by 15 degrees.",
        "cmd_byte": CMD_TURN_LEFT_15,
    },
    {
        "name": "turn_right_15",
        "description": "Turn the car right by 15 degrees.",
        "cmd_byte": CMD_TURN_RIGHT_15,
    },
    {
        "name": "move_forward_1ft",
        "description": "Move the car forwards by 1 foot.",
        "cmd_byte": CMD_MOVE_FWD_1FT,
    },
    {
        "name": "move_forward_2in",
        "description": "Move the car forwards by 2 inches.",
        "cmd_byte": CMD_MOVE_FWD_2IN,
    },
    {
        "name": "move_backward_1ft",
        "description": "Move the car backwards by 1 foot.",
        "cmd_byte": CMD_MOVE_BWD_1FT,
    },
    {
        "name": "move_backward_2in",
        "description": "Move the car backwards by 2 inches.",
        "cmd_byte": CMD_MOVE_BWD_2IN,
    },
]

# Build the Claude API tool schema from our definitions
TASK_COMPLETE_TOOL = {
    "name": "task_complete",
    "description": "Call this tool when you believe the task is fully complete. Provide a short summary of what was accomplished.",
    "input_schema": {
        "type": "object",
        "properties": {
            "summary": {
                "type": "string",
                "description": "A brief summary of what was accomplished.",
            },
        },
        "required": ["summary"],
    },
}


def get_claude_tools():
    """Return tool definitions in Claude API format."""
    movement_tools = [
        {
            "name": t["name"],
            "description": t["description"],
            "input_schema": {
                "type": "object",
                "properties": {},
                "required": [],
            },
        }
        for t in TOOL_DEFS
    ]
    return movement_tools + [TASK_COMPLETE_TOOL]

def get_cmd_byte(tool_name: str) -> int | None:
    """Map a tool name to its 1-byte command ID."""
    for t in TOOL_DEFS:
        if t["name"] == tool_name:
            return t["cmd_byte"]
    return None

File: ts-agent/src/tools.ts
// Command bytes (sent to Pi, followed by a 1-byte value)
export const CMD_BYTES: Record<string, number> = {
  turn_left: 0x01,
  turn_right: 0x02,
  move_forward: 0x03,
  move_backward: 0x04,
};

// Response bytes (from Pi)
export const RESP_ACK = 0xaa;
export const RESP_DONE = 0xbb;
export const RESP_ERROR = 0xff;

const REASONING_PARAM = {
  type: "string",
  description:
    "REQUIRED: Describe what you see in the current camera frame — where is the target relative to the green guide lines? Why are you making this move?",
};

// OpenAI Realtime function tool declarations
export const TOOL_DECLARATIONS = [
  {
    type: "function" as const,
    name: "turn_left",
    description:
      "Turn the car left by the specified number of degrees. Each unit is 15 degrees, so the value should be a multiple of 15.",
    parameters: {
      type: "object",
      properties: {
        reasoning: REASONING_PARAM,
        degrees: {
          type: "integer",
          description:
            "Number of degrees to turn left (must be a multiple of 15, e.g. 15, 30, 45, ... 180).",
        },
      },
      required: ["reasoning", "degrees"],
    },
  },
  {
    type: "function" as const,
    name: "turn_right",
    description:
      "Turn the car right by the specified number of degrees. Each unit is 15 degrees, so the value should be a multiple of 15.",
    parameters: {
      type: "object",
      properties: {
        reasoning: REASONING_PARAM,
        degrees: {
          type: "integer",
          description:
            "Number of degrees to turn right (must be a multiple of 15, e.g. 15, 30, 45, ... 180).",
        },
      },
      required: ["reasoning", "degrees"],
    },
  },
  {
    type: "function" as const,
    name: "move_forward",
    description:
      "Move the car forward by the specified number of inches. ONLY call this after calling look and confirming the target is between the green guide lines.",
    parameters: {
      type: "object",
      properties: {
        reasoning: REASONING_PARAM,
        inches: {
          type: "integer",
          description: "Number of inches to move forward (1-72).",
        },
      },
      required: ["reasoning", "inches"],
    },
  },
  {
    type: "function" as const,
    name: "move_backward",
    description: "Move the car backward by the specified number of inches.",
    parameters: {
      type: "object",
      properties: {
        reasoning: REASONING_PARAM,
        inches: {
          type: "integer",
          description: "Number of inches to move backward (1-72).",
        },
      },
      required: ["reasoning", "inches"],
    },
  },
  {
    type: "function" as const,
    name: "look",
    description:
      "Do nothing — just get a fresh camera frame. Use this to observe the scene before deciding on a movement. You should call this BEFORE calling move_forward to verify the target is between the green guide lines.",
    parameters: {
      type: "object",
      properties: {
        reasoning: {
          type: "string",
          description:
            "REQUIRED: Describe what you see in the current frame and what you are looking for. Where is the target? Is it between the green lines?",
        },
      },
      required: ["reasoning"],
    },
  },
  {
    type: "function" as const,
    name: "task_complete",
    description:
      "Call this ONLY when the task is fully complete (target is within arm's reach and fills most of the frame) or you are completely stuck and cannot make further progress. This ends the session.",
    parameters: {
      type: "object",
      properties: {
        reason: {
          type: "string",
          description:
            "Brief explanation of why the task is complete or why you cannot continue.",
        },
      },
      required: ["reason"],
    },
  },
];
